instrumentation/netty/netty-4-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/common/FutureListenerWrappers.java

```diff
@@ -44,21 +46,32 @@ public static boolean shouldWrap(GenericFutureListener<? extends Future<?>> list
   @SuppressWarnings("unchecked")
   public static GenericFutureListener<?> wrap(
       Context context, GenericFutureListener<? extends Future<?>> delegate) {
-    return wrappers.computeIfAbsent(
-        delegate,
-        key -> {
-          if (delegate instanceof GenericProgressiveFutureListener) {
-            return new WrappedProgressiveFutureListener(
-                context, (GenericProgressiveFutureListener<ProgressiveFuture<?>>) delegate);
-          } else {
-            return new WrappedFutureListener(context, (GenericFutureListener<Future<?>>) delegate);
-          }
-        });
+    WeakReference<GenericFutureListener<? extends Future<?>>> resultReference =
+        wrappers.computeIfAbsent(
+            delegate,
+            key -> {
+              GenericFutureListener<? extends Future<?>> wrapper;
+              if (delegate instanceof GenericProgressiveFutureListener) {
+                wrapper =
+                    new WrappedProgressiveFutureListener(
+                        context, (GenericProgressiveFutureListener<ProgressiveFuture<?>>) delegate);
+              } else {
+                wrapper =
+                    new WrappedFutureListener(context, (GenericFutureListener<Future<?>>) delegate);
+              }
+              return new WeakReference<>(wrapper);
+            });
+    return resultReference.get();
```

Comment:
we may need to be careful here, I think the cache can return a "cleared" weak value reference now (previously caffeine took care of not making "cleared" values visible).

Instead, I think we could use VirtualField instead of Cache, essentially reverting #3059 now that #4182 solves the underlying problem with VirtualField not being applied to lambdas.

Reply:
I sent two commits converting from Cache to VirtualField
* [Convert Netty wrapper cache to VirtualField](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4667/commits/dac1522a3f6fef096f379554e84e99f206229302)
* [Work around lambda instrumentation failure](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4667/commits/6d63815b44ca4721e9ee417f4f18314ee703bd80)

The second commit was needed to workaround a problem caused by the VirtualField instrumenting some of our own helper class lambdas. Normally I don't think helper classes are instrumented since they are injected from inside of ClassFileTransformer.transform (which is not recursive). But lambdas are different since they don't go through the normal class definition / transformation process.

I was already having second thoughts about the first commit above after typing up the warning about the memory leak that will occur if this code is ever used inside of library instrumentation.

I did add a [commit to handle cleared weak values](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4667/commits/9ab70f4d2aa52abcdb1a27ed917494cfe1276b69).